### PR TITLE
Add warning when accessing Tensor::grad() in the C++ API

### LINKED ISSUE #35379: Accessing Tensor::grad() in the C++ API for a non-leaf tensor should warn
## To Reproduce

In python, accessing the .grad attribute of a non-leaf tensor is a common pitfall users run into. For example:
```
x = torch.randn(3, requires_grad=True).cuda()
y = x ** 2
y.sum().backward()
x.grad
```
Luckily, this raises a nice warning in Python:
```
/scratch/rzou/pt/cudnn_double_Bwd/torch/tensor.py:746: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.
  warnings.warn("The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad "
```

In the C++ API, no warning happens: https://discuss.pytorch.org/t/very-confused-with-changes-in-autograd/74355 . It would be great for usability to add that warning.

cc @ezyang @SsnL @albanD @zou3519 @gqchen @yf225

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #35379

 - Adds  `retains_grad` attribute backed by cpp as a native function. The python bindings for the function are skipped to be consistent with `is_leaf`.
   - Tried writing it without native function, but the jit test `test_tensor_properties` seems to require that it be a native function (or alternatively maybe it could also work if we manually add a prim implementation?).
 - Python API now uses `retain_grad` implementation from cpp

### CODE REVIEW

File: aten/src/ATen/core/VariableHooksInterface.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ struct TORCH_API VariableHooksInterface {
   virtual Tensor data(const Tensor&) const = 0;
   virtual int64_t _version(const Tensor&) const = 0;
   virtual void retain_grad(const Tensor&) const = 0;
+  virtual bool _is_retain_grad(const Tensor&) const = 0;
```
Comment (albanD) [inline aten/src/ATen/core/VariableHooksInterface.h:57]: nit: why not use the same name as the python API: `retains_grad` ?

File: aten/src/ATen/templates/TensorBody.h [hunk 2]
```diff
@@ -706,7 +706,16 @@ class TORCH_API Tensor {
   /// The attribute will then contain the gradients computed and future calls
   /// to `backward()` will accumulate (add) gradients into it.
   const Tensor& grad() const {
-    return impl_->grad();
+    const Tensor& maybe_grad = impl_->grad();
+    if (!is_leaf() && !_is_retain_grad() && !maybe_grad.defined()) {
+      TORCH_WARN(
+        "The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad "
+        "attribute won't be populated during autograd.backward(). If you indeed want the gradient "
```
Comment (albanD) [inline aten/src/ATen/templates/TensorBody.h:713]: nit: `want the gradient` -> `want the .grad field to be populated`.

File: torch/_tensor.py [hunk 3]
```diff
@@ -988,12 +988,6 @@ def grad(self):
             # TODO mypy doesn't support @property, see: https://github.com/python/mypy/issues/6185
             return handle_torch_function(Tensor.grad.__get__, (self,), self)  # type: ignore[attr-defined]
 
-        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf and self._grad is None:
```
Comment (albanD) [inline torch/_tensor.py:991]: Should you remove all the logic related to `retains_grad` here?

Comment (albanD) [reply torch/_tensor.py:991]: Or make it an attribute backed by the c++ implementation

File: torch/csrc/autograd/variable.cpp [hunk 4]
```diff
@@ -376,23 +377,23 @@ Tensor VariableHooks::tensor_data(const Tensor& self) const {
   return at::Tensor(self_impl_copy);
 }
 
-bool VariableHooks::is_leaf(const Tensor & self) const {
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:379]: nit: do codemod in another PR :D

**Review by albanD — COMMENTED** [review]
Is there a reason why we can't just delete the python implementation of retain_grad and bind the c++ one? The two do the same thing no?

File: torch/csrc/autograd/python_variable.cpp [hunk 5]
```diff
@@ -504,6 +504,16 @@ PyObject *THPVariable_get_requires_grad(THPVariable *self, void *unused)
   END_HANDLE_TH_ERRORS
 }
 
+PyObject *THPVariable_retains_grad(THPVariable *self, void *unused)
+{
+  HANDLE_TH_ERRORS
+  if (check_has_torch_function((PyObject *)self)) {
+    return handle_torch_function_getter(self, "retains_grad");
+  }
+  return PyBool_FromLong(THPVariable_Unpack(self).retains_grad());
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:513]: It's surprising to convert form Long here given that retains_grad is already a bool.
You might want to check the bool in c++ and then use `Py_RETURN_TRUE` and `Py_RETURN_FALSE`.

File: aten/src/ATen/templates/TensorBody.h [hunk 6]
```diff
@@ -706,7 +712,16 @@ class TORCH_API Tensor {
   /// The attribute will then contain the gradients computed and future calls
   /// to `backward()` will accumulate (add) gradients into it.
   const Tensor& grad() const {
-    return impl_->grad();
+    const Tensor& maybe_grad = impl_->grad();
+    if (!is_leaf() && !retains_grad() && !maybe_grad.defined()) {
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply aten/src/ATen/templates/TensorBody.h:716]: I don't think we need that, since non-leaf tensors are automatically requires_grad==True

File: tools/autograd/gen_python_functions.py [hunk 7]
```diff
@@ -95,7 +95,7 @@
     'nonzero(_(out|numpy))?',
     'set_data',
     '.*_overrideable',  # overrideable functions for backend extension
-    'data', 'is_leaf', 'output_nr', '_version', 'requires_grad_', 'retain_grad', 'set_',
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply tools/autograd/gen_python_functions.py:98]: Its removed from this list so that the python bindings would be generated, as its a native function. I'm guessing it was here previously because we already had the duplicate python version

File: aten/src/ATen/templates/TensorBody.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply aten/src/ATen/templates/TensorBody.h:716]: I don't think we need that, since non-leaf implies requires_grad==True, so we also only warn when requires_grad=True

File: torch/csrc/autograd/python_variable.cpp (hunk 5 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_variable.cpp:513]: Yeah that is interesting, fixed.

File: aten/src/ATen/templates/TensorBody.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (albanD) [reply aten/src/ATen/templates/TensorBody.h:716]: Good point.

**Review by albanD — APPROVED** [review]
LGTM

Lint failures are real though, you should fix that before landing.
